web/packages/teleport/src/Roles/RoleEditor/StandardEditor.tsx

```diff
@@ -896,6 +927,84 @@ export function WindowsDesktopAccessSpecSection({
   );
 }
 
+export function AdminRules({
+  value,
+  isProcessing,
+  validation,
+  onChange,
+}: SectionProps<RuleModel[], AdminRuleValidationResult[]>) {
+  function addRule() {
+    onChange?.([...value, newRuleModel()]);
+  }
+  function setRule(rule: RuleModel) {
+    onChange?.(value.map(r => (r.id === rule.id ? rule : r)));
+  }
+  function removeRule(id: string) {
+    onChange?.(value.filter(r => r.id !== id));
+  }
+  return (
+    <Flex flexDirection="column" gap={3}>
+      {value.map((rule, i) => (
+        <AdminRule
+          key={rule.id}
+          isProcessing={isProcessing}
+          value={rule}
+          onChange={setRule}
+          validation={validation[i]}
+          onRemove={() => removeRule(rule.id)}
+        />
+      ))}
+      <ButtonSecondary alignSelf="start" onClick={addRule}>
+        <Icon.Plus size="small" mr={2} />
+        Add New
+      </ButtonSecondary>
+    </Flex>
+  );
+}
+
+function AdminRule({
+  value,
+  isProcessing,
+  validation,
+  onChange,
+  onRemove,
+}: SectionProps<RuleModel, AdminRuleValidationResult> & {
+  onRemove?(): void;
+}) {
+  const { resources, verbs } = value;
+  const theme = useTheme();
+  return (
+    <Section
+      title="Admin Rule"
+      tooltip="A rule that gives users administrative access to certain kinds of resources"
+      removable
+      isProcessing={isProcessing}
+      validation={validation}
```

Comment:
It's a minor issue, but when I click "Create role" while there's an invalid admin rule in the form, any other new role that I add gets automatically validated and highlighted as invalid, since I'm yet to fill out any fields.

I don't think it's a blocker, personally I just don't like forms that tell me that a field is invalid before I had a chance to even touch it. ;)

https://github.com/user-attachments/assets/8056abc3-5f14-459c-acad-fb140fef6043

Reply:
@ravicious Unfortunately, our validation library doesn't support a notion of a touched input. It only supports a global "validating" state. I may try to extend it in future, but to be honest, I think that we'd better off just migrating to a third-party library instead.